common/go-utils/src/inv.rs

```diff
@@ -70,19 +62,61 @@ struct GoRelease {
 struct GoFile {
     os: String,
     arch: String,
+    filename: String,
     sha256: String,
     version: String,
 }
 
-impl GoFile {
-    fn target_arch(&self) -> Option<String> {
-        if self.os.is_empty() || self.arch.is_empty() {
-            return None;
-        }
-        Some(format!("{}-{}", self.os, self.arch))
+#[derive(thiserror::Error, Debug)]
+pub enum GoFileConversionError {
+    #[error(transparent)]
+    Version(#[from] VersionParseError),
+    #[error("Arch is not supported: {0}")]
+    Arch(String),
+    #[error("OS is not supported: {0}")]
+    Os(String),
+}
+
+impl TryFrom<&GoFile> for Artifact {
+    type Error = GoFileConversionError;
+
+    fn try_from(value: &GoFile) -> Result<Self, Self::Error> {
+        Ok(Artifact {
+            go_version: value.version.clone(),
+            semantic_version: Version::parse_go(&value.version)?,
+            os: parse_go_os(&value.os)?,
+            arch: parse_go_arch(&value.arch)?,
```

Comment:
@runesoerensen this is a good example of where you could go beyond a `String` value.  For example:

```rust
enum GoArch {
  x86_64,
  aarch64,
}
```

And to get the conversion from a string to your enum, implement [FromStr](https://doc.rust-lang.org/std/str/trait.FromStr.html):

```rust
struct UnsupportedGoArchError(String);

impl FromStr for GoArch {
  type Err = UnsupportedGoArchError;

  fn from_str(s: &str) -> Result<Self, Self::Err> {
    match s {
        "amd64" => Ok(GoArch::x86_64),
        "arm64" => Ok(GoArch::aarch64),
        _ => Err(UnsupportedGoArchError(arch.to_string())),
    }
  }
}
```

And you'd use that like so:

```rust
Ok(Artifact {
    // ...
    arch: &value.arch.parse::<GoArch>()?,
})
```

I wouldn't say it's required for this PR but the idea would apply because:
* you've got a constrained set of acceptable `String` values you want to support (`amd64`, `arm64`)
* you're parsing the value from a `String` to ensure it's only one of those values but then returning the equally broad type of `String` instead of something that indicates it's a value in the acceptable set

Reply:
Definitely hear you on this -- perhaps this is something I can do in a follow-up PR?

Reply:
Fair point -- the reason I wanted to do it at a later time was mostly because I didn't want to introduce a Go-specific `GoArch` on `Artifact`, but rather try and use a more generic `Arch` enum. That's mostly semantics really, but basically the Arch parsing will need to support non-go specific types for this to work (in particular, the values we will store and Rust uses for constants etc).

So it was a bit more involved I guess - I've pushed some changes that I think will maybe address this (and your comment regarding the OS as well)?